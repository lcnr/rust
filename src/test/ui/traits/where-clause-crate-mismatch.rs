// aux-build:crate_a1.rs
// aux-build:crate_a2.rs

// FIXME(#71954) Paths to extern crates contain the scope of their first appearance.
// If we were to remove this function, `crate_a2` would have the path `inner::a::Bar`
// while `crate_a1` would use `a::Bar`.
pub fn hack() {
    {
        extern crate crate_a1 as a;
    }
    {
        extern crate crate_a2 as a;
    }
}



mod inner {
    extern crate crate_a2 as a;
    pub struct B<T: a::Bar>(T);
    //~^ NOTE required by this bound
    //~| NOTE required by this bound
}

extern crate crate_a1 as a;
pub struct C<T: a::Bar>(inner::B<T>);
//~^ ERROR the trait bound
//~| HELP consider further restricting
//~| NOTE not implemented for `T`
//~| NOTE mismatched crates

trait Bar {}

// Do not warn if the traits have the same name but
// differently named crates.
pub struct D<T: Bar>(inner::B<T>);
//~^ ERROR the trait bound
//~| HELP consider further restricting
//~| NOTE not implemented for `T`

fn main() {}
