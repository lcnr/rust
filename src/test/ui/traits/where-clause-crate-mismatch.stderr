error[E0277]: the trait bound `T: hack::a::Bar` is not satisfied
  --> $DIR/where-clause-crate-mismatch.rs:26:25
   |
LL |     pub struct B<T: a::Bar>(T);
   |                     ------ required by this bound in `inner::B`
...
LL | pub struct C<T: a::Bar>(inner::B<T>);
   |                         ^^^^^^^^^^^ the trait `hack::a::Bar` is not implemented for `T`
   |
help: consider further restricting this bound
   |
LL | pub struct C<T: a::Bar + hack::a::Bar>(inner::B<T>);
   |                        ^^^^^^^^^^^^^^

error[E0277]: the trait bound `T: hack::a::Bar` is not satisfied
  --> $DIR/where-clause-crate-mismatch.rs:36:22
   |
LL |     pub struct B<T: a::Bar>(T);
   |                     ------ required by this bound in `inner::B`
...
LL | pub struct D<T: Bar>(inner::B<T>);
   |                      ^^^^^^^^^^^ the trait `hack::a::Bar` is not implemented for `T`
   |
help: consider further restricting this bound
   |
LL | pub struct D<T: Bar + hack::a::Bar>(inner::B<T>);
   |                     ^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
